containerd: properly populate /etc/hosts and /etc/hostname

muntac opened the pull request:
<!--
Thanks for submitting a pull request to Concourse!

If you haven't already, feel free to [add yourself] as a contributor so that
you can add labels to your PR and re-trigger its builds if they fail.

Also check the [PR requirements] if you haven't already!
-->

[add yourself]: https://github.com/concourse/governance#individual-contributors
[PR requirements]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md#pull-request-requirements

## Changes proposed by this PR
cc: @chenbh 

closes #6811

<!--
Summarize your changes as a checklist, leaving any unfinished work as unchecked
items. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->
* [x] add hostname to `/etc/hostname`
* [x] add container's IP and hostname to `/etc/hosts`

## Notes to reviewer

<!--
If needed, leave any special pointers for reviewing or testing your PR.
-->
The container's IP is only available once the network has been initialized. Therefore, the original /etc/hosts has to be updated after. The /etc/hosts on the container is bindmounted to a file on the worker.  We simply update this
file directly.

To test simply jump on to a running container and open the `/etc/hosts` and `/etc/hostname` file. You can also run `hostname -f` and `hostname -i` to validate.

## Release Note

<!--
Your PR title will be directly included in the release notes when it ships in
the next Concourse release. It should briefly describe the PR in [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Within this section you may supply a list of extra information to include in
the release notes in addition to the pull request title.

Example title: Introduce new pipeline UI algorithm

Example notes:

* Reticulating splines is the new process Concourse uses to create the network
  of lines between jobs.
* Combines many short lines and curves into a network of splines.

If there are no additional notes necessary you may remove this entire section.
-->

* containerd: `/etc/hosts` and `/etc/hostname` are correctly populated

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

Commit:
containerd: add hostname to appropriate files

The hostname is the container's handle. This is now added to /etc/hostname.

Both the hostname and the container's IP is appended to /etc/hosts. The
container's IP is only available once the network has been initialized.
Therefore, the original /etc/hosts has to be updated after. The /etc/hosts on
the container is bindmounted to a file on the worker.  We simply update this
file directly.

issue#6811

Co-authored-by: Muntasir Chowdhury <mchowdhury@pivotal.io>
Signed-off-by: Bohan Chen <bochen@pivotal.io>

taylorsilva commented on worker/runtime/integration/integration_test.go:
```diff
@@ -434,6 +434,65 @@ func (s *IntegrationSuite) TestContainerAllowsHostAccess() {
 }
 
+func (s *IntegrationSuite) TestContainerNetworkHosts() {
+	namespace := "test-restricted-networks"
+	requestTimeout := 3 * time.Second
+
+	network, err := runtime.NewCNINetwork()
+
+	s.NoError(err)
+
+	networkOpt := runtime.WithNetwork(network)
+	customBackend, err := runtime.NewGardenBackend(
+		libcontainerd.New(
+			s.containerdSocket(),
+			namespace,
+			requestTimeout,
+		),
+		networkOpt,
+	)
+	s.NoError(err)
+
+	s.NoError(customBackend.Start())
+
```
We don't need to create a custom garden backend if we're not changing anything from the default values. Use `s.gardenBackend` instead.
```suggestion
```

taylorsilva commented on worker/runtime/integration/integration_test.go:
```diff
@@ -434,6 +434,65 @@ func (s *IntegrationSuite) TestContainerAllowsHostAccess() {
+
+	handle := uuid()
+
+	container, err := customBackend.Create(garden.ContainerSpec{
```
```suggestion
	container, err := s.gardenBackend.Create(garden.ContainerSpec{
```

taylorsilva commented on worker/runtime/integration/integration_test.go:
```diff
@@ -434,6 +434,65 @@ func (s *IntegrationSuite) TestContainerAllowsHostAccess() {
+	s.NoError(err)
+
+	defer func() {
+		s.NoError(customBackend.Destroy(handle))
+		customBackend.Stop()
```
```suggestion
		s.NoError(s.gardenBackend.Destroy(handle))
```

taylorsilva commented on worker/runtime/integration/integration_test.go:
```diff
@@ -434,6 +434,65 @@ func (s *IntegrationSuite) TestContainerAllowsHostAccess() {
+	exitCode, err := proc.Wait()
+	s.NoError(err)
+
+	fmt.Println(buf.String())
```
from debugging I'm guessing
```suggestion
```

muntac committed:
containerd: clean up network host integration test

Use already provided s.GardenBackend instead of creating custom backend.
Remove println used for debugging.

Signed-off-by: Muntasir Chowdhury <mchowdhury@pivotal.io>
Signed-off-by: Muntasir Chowdhury <muntasir.mzc@gmail.com>

taylorsilva reviewed (approved):
LGTM

muntac commented:
Thanks! Good eye with the test cleanup suggestions.

jaredstehler commented:
would it be possible to get this included in the next point release?

steveburton4 commented:
@taylorsilva Do you know when this is planned to be released?

taylorsilva commented:
Soon maybe? Looking at our milestone: https://github.com/concourse/concourse/milestone/75
It looks like we could probably cut another release soon.

Christian-Oleson commented:
@jaredstehler , 
If I am reading this right it looks like this fix went out in the last release:
https://github.com/concourse/concourse/releases/tag/v7.4.0

steveburton4 commented:
> @jaredstehler ,
> If I am reading this right it looks like this fix went out in the last release:
> https://github.com/concourse/concourse/releases/tag/v7.4.0

It did. We have installed it already and tested it fixes the issue we were having 👍
